acme/acme/messages.py

```diff
@@ -184,7 +185,7 @@ class Identifier(jose.JSONObjectWithFields):
 class Directory(jose.JSONDeSerializable):
     """Directory."""
 
-    _REGISTERED_TYPES: dict = {}
+    _REGISTERED_TYPES: Dict = {}
```

Comment:
Well, ideally I think it should be `Dict[str, ResourceMixin]` or `Dict[str, ResourceBody]`, but none of them declares the required class-level field `resource_type`. It is just dynamically declared on all subclasses of `ResourceBody` that are actually using the decorator `Directory.register`.

Adding it to the mixin would require in fact to declare a method that allows you to retrieve the resource type, but we will need to use some advanced mypy techniques like `Protocol` to make it happy.

We could also add the class-level property to `ResourceBody`, but I fear that it will not play nice with the parent class `JSONObjectWithFields`, or could have undesirable side-effects for classes subclassing `ResourceBody` but do not need the `resource-type` property.

At the minimum we can declare `Dict[str, Type[Any]]`, it is not great, but at least we know that this is a dict mapping strings keys to class definitions. I did that.